mujoco_warp/_src/derivative.py

```diff
@@ -210,6 +211,496 @@ def _qderiv_tendon_damping(
       qDeriv_out[worldid, dofjid, dofiid] -= qderiv
 
 
+@wp.func
+def _pow2_deriv(val: float) -> float:
+  return val * val
+
+
+@wp.func
+def _geom_semiaxes_deriv(size: wp.vec3, geom_type: int) -> wp.vec3:
+  if geom_type == GeomType.SPHERE:
+    r = size[0]
+    return wp.vec3(r, r, r)
+
+  if geom_type == GeomType.CAPSULE:
+    radius = size[0]
+    half_length = size[1]
+    return wp.vec3(radius, radius, half_length + radius)
+
+  if geom_type == GeomType.CYLINDER:
+    radius = size[0]
+    half_length = size[1]
+    return wp.vec3(radius, radius, half_length)
+
+  return size
+
+
+@wp.func
+def _ellipsoid_max_moment_deriv(size: wp.vec3, dir: int) -> float:
+  d0 = size[dir]
+  d1 = size[(dir + 1) % 3]
+  d2 = size[(dir + 2) % 3]
+  d_max = wp.max(d1, d2)
+  return wp.static(8.0 / 15.0 * wp.pi) * d0 * d_max * d_max * d_max * d_max
+
+
+@wp.func
+def _add_to_quadrant(B: wp.mat66, D: wp.mat33, col_quad: int, row_quad: int) -> wp.mat66:
+  r = 3 * row_quad
+  c = 3 * col_quad
+  for i in range(3):
+    for j in range(3):
+      B[r + i, c + j] += D[i, j]
+  return B
+
+
+@wp.func
+def _cross_deriv_a(a: wp.vec3, b: wp.vec3) -> wp.mat33:
```

Comment:
lets consider combining `_cross_deriv_a` and `_cross_deriv_b` into one function that returns 2 `wp.mat33` since it looks like we always compute derivatives wrt to both cross product inputs